src/model/ImageDocument.cpp

```diff
@@ -140,18 +235,41 @@ QVector4D ImageDocument::pixelValue() const
   return texel;
 }
 
-bool ImageDocument::check(Result<bool> const& result)
+void ImageDocument::load(QString const& path, QFutureWatcher<LoadResult>* watcher)
 {
-  if (!result) {
-    setError(QString::fromStdString(result.error()));
-  }
-  return (bool)result;
+  QFuture<LoadResult> future = QtConcurrent::run([path]() {
+    QFileInfo file(path);
+    if (!file.exists()) {
+      return std::make_shared<Result<Image>>(QString("File " + file.absoluteFilePath() + " does not exist.").toStdString());
+    }
+    if (file.suffix() == "hdr" || file.suffix() == "pic") {
+      return std::make_shared<Result<Image>>(Image::loadPIC(file.absoluteFilePath().toStdString()));
+    } else if (file.suffix() == "pfm" || file.suffix() == "ppm") {
+      return std::make_shared<Result<Image>>(Image::loadPFM(file.absoluteFilePath().toStdString()));
+    } else if (file.suffix() == "exr") {
+      return std::make_shared<Result<Image>>(Image::loadEXR(file.absoluteFilePath().toStdString()));
+    } else {
+      return std::make_shared<Result<Image>>(Image::loadImage(file.absoluteFilePath().toStdString()));
+    }
+    return std::make_shared<Result<Image>>(Result<Image>(QString("Unknown file extension: " + file.suffix()).toStdString()));
+  });
+  watcher->setFuture(future);
 }
 
-void ImageDocument::setError(QString const& message)
+void ImageDocument::loadFinished(QFutureWatcher<LoadResult>* watcher, QUrl const& url, bool comparison)
 {
-  auto & collection = *static_cast<ImageCollection *>(parent());
-  collection.setError(message);
+  auto & result = watcher->result();
+  if (check(*result, comparison ? ErrorCategory::E_Comparsion : ErrorCategory::E_Image, "Failed to load " + url.toLocalFile() + ": ")) {
+    if (!comparison) {
+      image_ = std::make_shared<Image>(result->value());
```

Comment:
This will copy the whole Image I think? Do `std::move(*result).value()` to move. It's a bit dangerous because technically we could still access watcher->result() a second time (is there a way to reset the watcher?)